src/dotnet/ShellShim/ShellShimManager.cs

```diff
@@ -0,0 +1,213 @@
+                        }
+                        else
+                        {
+                            var script = new StringBuilder();
+                            script.AppendLine("#!/bin/sh");
+                            script.AppendLine($"dotnet {targetExecutablePath.ToQuotedString()} \"$@\"");
+
+                            var shimPath = GetPosixShimPath(commandName);
+                            File.WriteAllText(shimPath.Value, script.ToString());
+
+                            SetUserExecutionPermission(shimPath);
+                        }
+                    },
+                    rollback: () => {
+                        foreach (var file in GetShimFiles(commandName).Where(f => File.Exists(f.Value)))
+                        {
+                            File.Delete(file.Value);
+                        }
+                    });
+            }
+            catch (Exception ex) when (ex is UnauthorizedAccessException || ex is IOException)
+            {
+                throw new ShellShimException(
+                    string.Format(
+                        CommonLocalizableStrings.FailedToCreateShellShim,
+                        commandName,
+                        ex.Message
+                    ),
+                    ex);
+            }
+        }
+
+        public void RemoveShim(string commandName)
+        {
+            try
+            {
+                var files = new Dictionary<string, string>();
+                TransactionalAction.Run(
+                    action: () => {
+                        foreach (var file in GetShimFiles(commandName).Where(f => File.Exists(f.Value)))
+                        {
+                            var tempPath = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());
+                            File.Move(file.Value, tempPath);
+                            files[file.Value] = tempPath;
+                        }
+                    },
+                    commit: () => {
+                        foreach (var value in files.Values)
+                        {
+                            File.Delete(value);
+                        }
+                    },
+                    rollback: () => {
+                        foreach (var kvp in files)
+                        {
+                            File.Move(kvp.Value, kvp.Key);
+                        }
+                    });
+            }
+            catch (Exception ex) when (ex is UnauthorizedAccessException || ex is IOException)
+            {
+                throw new ShellShimException(
+                    string.Format(
+                        CommonLocalizableStrings.FailedToRemoveShellShim,
+                        commandName,
+                        ex.Message
+                    ),
+                    ex);
+            }
+        }
+
+        public bool ShimExists(string commandName)
+        {
+            return GetShimFiles(commandName).Any(p => File.Exists(p.Value));
```

Comment:
Also the config file? how do you make sure config file and exe will exist or delete at the same time

Reply:
This will check the config file on Windows as well (the iterator returns both the `.exe` and `.exe.config`).